kolibri/auth/test/test_collections_and_roles.py

```diff
@@ -115,34 +122,89 @@ class CollectionsTestCase(TestCase):
+        self.classroom.remove_admin(user)
+        self.facility.remove_admin(user)
+        self.assertEqual(Role.objects.filter(user=user, kind=Role.KIND_ADMIN, collection=self.classroom).count(), 0)
+        self.assertEqual(Role.objects.filter(user=user, kind=Role.KIND_ADMIN, collection=self.facility).count(), 0)
 
-    def test_add_classroom(self):
-        Classroom.objects.create(parent=self.facility)
-        self.assertEqual(Classroom.objects.count(), 1)
-
-    def test_add_coach(self):
+    def test_add_and_remove_coach(self):
         user = FacilityUser.objects.create(username='foo', facility=self.facility)
+        self.classroom.add_coach(user)
+        self.facility.add_coach(user)
+        self.assertEqual(Role.objects.filter(user=user, kind=Role.KIND_COACH, collection=self.classroom).count(), 1)
+        self.assertEqual(Role.objects.filter(user=user, kind=Role.KIND_COACH, collection=self.facility).count(), 1)
+        self.classroom.remove_coach(user)
+        self.facility.remove_coach(user)
+        self.assertEqual(Role.objects.filter(user=user, kind=Role.KIND_COACH, collection=self.classroom).count(), 0)
+        self.assertEqual(Role.objects.filter(user=user, kind=Role.KIND_COACH, collection=self.facility).count(), 0)
+
+    def test_add_coaches(self):
+        user1 = FacilityUser.objects.create(username='foo1', facility=self.facility)
+        user2 = FacilityUser.objects.create(username='foo2', facility=self.facility)
+        self.classroom.add_coaches([user1, user2])
+        self.facility.add_coaches([user1, user2])
+        self.assertEqual(Role.objects.filter(kind=Role.KIND_COACH, collection=self.classroom).count(), 2)
+        self.assertEqual(Role.objects.filter(kind=Role.KIND_COACH, collection=self.facility).count(), 2)
+
+    def test_add_admins(self):
+        user1 = FacilityUser.objects.create(username='foo1', facility=self.facility)
+        user2 = FacilityUser.objects.create(username='foo2', facility=self.facility)
+        self.classroom.add_admins([user1, user2])
+        self.facility.add_admins([user1, user2])
+        self.assertEqual(Role.objects.filter(kind=Role.KIND_ADMIN, collection=self.classroom).count(), 2)
+        self.assertEqual(Role.objects.filter(kind=Role.KIND_ADMIN, collection=self.facility).count(), 2)
+
+    def test_add_classroom(self):
         classroom = Classroom.objects.create(parent=self.facility)
-        classroom.add_coach(user)
-        self.assertEqual(Role.objects.filter(user=user, kind=Role.KIND_COACH, collection=classroom).count(), 1)
+        self.assertEqual(Classroom.objects.count(), 2)
+        self.assertEqual(classroom.get_facility(), self.facility)
 
     def test_add_learner_group(self):
-        classroom = Classroom.objects.create(parent=self.facility)
+        classroom = Classroom.objects.create(name="blah", parent=self.facility)
+        classroom.full_clean()
```

Comment:
It gets called automatically by forms, and will be called by Morango when bringing in synced data.

A dev doesn't _need_ to call it at a particular time, but it can be called if validation is desired (beyond the basic DB integrity validation that `save` does). Ideally, any code calling `full_clean` would handle `ValidationError`s that might arise.